src/main/java/de/tum/in/www1/artemis/service/QuizService.java

```diff
@@ -0,0 +1,274 @@
+                            dropLocationCounter.setDropLocation(null);
+                            dropLocationCounterToDelete.add(dropLocationCounter);
+                        }
+                    }
+                }
+                quizQuestionStatistic.getDropLocationCounters().removeAll(dropLocationCounterToDelete);
+
+                // save references as index to prevent Hibernate Persistence problem
+                saveCorrectMappingsInIndices(dndQuestion);
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion saQuestion) {
+                var quizQuestionStatistic = (ShortAnswerQuestionStatistic) saQuestion.getQuizQuestionStatistic();
+                if (quizQuestionStatistic == null) {
+                    quizQuestionStatistic = new ShortAnswerQuestionStatistic();
+                    saQuestion.setQuizQuestionStatistic(quizQuestionStatistic);
+                    quizQuestionStatistic.setQuizQuestion(quizQuestion);
+                }
+
+                for (var spot : saQuestion.getSpots()) {
+                    spot.setQuestion(saQuestion);
+                    quizQuestionStatistic.addSpot(spot);
+                }
+
+                // if a spot was removed then remove the associated spotCounters implicitly
+                Set<ShortAnswerSpotCounter> spotCounterToDelete = new HashSet<>();
+                for (ShortAnswerSpotCounter spotCounter : quizQuestionStatistic.getShortAnswerSpotCounters()) {
+                    if (spotCounter.getId() != null) {
+                        if (!(saQuestion.getSpots().contains(spotCounter.getSpot()))) {
+                            spotCounter.setSpot(null);
+                            spotCounterToDelete.add(spotCounter);
+                        }
+                    }
+                }
+                quizQuestionStatistic.getShortAnswerSpotCounters().removeAll(spotCounterToDelete);
+
+                // save references as index to prevent Hibernate Persistence problem
+                saveCorrectMappingsInIndicesShortAnswer(saQuestion);
+            }
+        }
+
+        T savedQuizConfiguration = saveAndFlush(quizConfiguration);
+
+        // fix references in all drag and drop questions and short answer questions (step 2/2)
+        for (QuizQuestion quizQuestion : savedQuizConfiguration.getQuizQuestions()) {
+            if (quizQuestion instanceof DragAndDropQuestion dragAndDropQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndices(dragAndDropQuestion);
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion shortAnswerQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesShortAnswer(shortAnswerQuestion);
+            }
+        }
+
+        return savedQuizConfiguration;
+    }
+
```

Comment:
Please stay consistent with your variable declarations and refactor each instanceof case into its own function. The for loops in line 57 and 81 and 109 are the same code, you can refactor it into a function with a generic parameter.

Reply:
Codes are moved to separate functions for each case and generic method findTobeDeletedComponent is created